README.md
<br>English | [简体中文](README_zh.md)

[![CI](https://github.com/gngpp/opengpt/actions/workflows/CI.yml/badge.svg)](https://github.com/gngpp/opengpt/actions/workflows/CI.yml)
[![CI](https://github.com/gngpp/opengpt/actions/workflows/Release.yml/badge.svg)](https://github.com/gngpp/opengpt/actions/workflows/Release.yml)
	<a target="_blank" href="https://github.com/gngpp/vdns/blob/main/LICENSE">
		<img src="https://img.shields.io/badge/license-MIT-blue.svg"/>
	</a>
  <a href="https://github.com/gngpp/opengpt/releases">
    <img src="https://img.shields.io/github/release/gngpp/opengpt.svg?style=flat">
  </a><a href="https://github.com/gngpp/opengpt/releases">
    <img src="https://img.shields.io/github/downloads/gngpp/opengpt/total?style=flat">
  </a>
  [![](https://img.shields.io/docker/image-size/gngpp/opengpt)](https://registry.hub.docker.com/r/gngpp/opengpt)
  [![Docker Image](https://img.shields.io/docker/pulls/gngpp/opengpt.svg)](https://hub.docker.com/r/gngpp/opengpt/)

# opengpt

A reverse engineered unofficial `ChatGPT` proxy (bypass Cloudflare 403 Access Denied)

- API key acquisition
- Email/password account authentication (Google/Microsoft third-party login is not supported for now because the author does not have an account)
- Unofficial/Official Http API proxy (for third-party client access)
- The original ChatGPT WebUI

> Limitations: This cannot bypass OpenAI's outright IP ban

### Platform Support

- Linux musl current supports
  - `x86_64-unknown-linux-musl`
  - `aarch64-unknown-linux-musl`
  - `armv7-unknown-linux-musleabi`
  - `armv7-unknown-linux-musleabihf`
  - `arm-unknown-linux-musleabi`
  - `arm-unknown-linux-musleabihf`
  - `armv5te-unknown-linux-musleabi`
- Windows current supports
  - `x86_64-pc-windows-msvc`
- MacOS current supports
  - `x86_64-apple-darwin`
  - `aarch64-apple-darwin`

### Install

  > #### Ubuntu(Other Linux)

Making [Releases](https://github.com/gngpp/opengpt/releases/latest) has a precompiled deb package, binaries, in Ubuntu, for example:

```shell
wget https://github.com/gngpp/opengpt/releases/download/v0.2.5/opengpt-0.2.5-x86_64-unknown-linux-musl.deb

dpkg -i opengpt-0.2.5-x86_64-unknown-linux-musl.deb

opengpt serve run
```

> #### Docker

```shell
docker run --rm -it -p 7999:7999 --hostname=opengpt \
  -e OPENGPT_WORKERS=1 \
  -e OPENGPT_LOG_LEVEL=info \
  gngpp/opengpt:latest serve run
```

> #### OpenWrt

There are pre-compiled ipk files in GitHub [Releases](https://github.com/gngpp/opengpt/releases/latest), which currently provide versions of aarch64/x86_64 and other architectures. After downloading, use opkg to install, and use nanopi r4s as example:

```shell
wget https://github.com/gngpp/opengpt/releases/download/v0.2.5/opengpt_0.2.5_aarch64_generic.ipk
wget https://github.com/gngpp/opengpt/releases/download/v0.2.5/luci-app-opengpt_1.0.2-1_all.ipk
wget https://github.com/gngpp/opengpt/releases/download/v0.2.5/luci-i18n-opengpt-zh-cn_1.0.2-1_all.ipk

opkg install opengpt_0.2.5_aarch64_generic.ipk
opkg install luci-app-opengpt_1.0.2-1_all.ipk
opkg install luci-i18n-opengpt-zh-cn_1.0.2-1_all.ipk
```

### Command Line(dev)

### Http Server

- Authentic ChatGPT WebUI
- Support unofficial/official API proxy
- The API prefix is consistent with the official
- Accessible to third-party clients
- Access to IP proxy pool to increase concurrency
- API documentation

- Parameter Description
  - Platfrom API [doc](https://platform.openai.com/docs/api-reference)
  - Backend API [doc](doc/rest.http)

- Parameter Description
  - `--level`, environment variable `OPENGPT_LOG_LEVEL`, log level: default info
  - `--host`, environment variable `OPENGPT_HOST`, service listening address: default 0.0.0.0,
  - `--port`, environment variable `OPENGPT_PORT`, listening port: default 7999
  - `--workers`, environment variable `OPENGPT_WORKERS`, worker threads: default 1
  - `--tls-cert`, environment variable `OPENGPT_TLS_CERT`', TLS certificate public key. Supported format: EC/PKCS8/RSA
  - `--tls-key`, environment variable `OPENGPT_TLS_KEY`, TLS certificate private key
  - `--proxies`, environment variable `OPENGPT_PROXY`, proxies，support multiple proxy pools, format: protocol://user:pass@ip:port

```shell
$ opengpt serve --help
Start the http server

Usage: opengpt serve run [OPTIONS]

Options:
  -H, --host <HOST>
          Server Listen host [env: OPENGPT_HOST=] [default: 0.0.0.0]
  -P, --port <PORT>
          Server Listen port [env: OPENGPT_PORT=] [default: 7999]
  -W, --workers <WORKERS>
          Server worker-pool size (Recommended number of CPU cores) [env: OPENGPT_WORKERS=] [default: 1]
  -L, --level <LEVEL>
          Log level (info/debug/warn/trace/error) [env: OPENGPT_LOG_LEVEL=] [default: info]
      --proxies <PROXIES>
          Server proxies pool, example: protocol://user:pass@ip:port [env: OPENGPT_PROXY=]
      --tcp-keepalive <TCP_KEEPALIVE>
          TCP keepalive (second) [env: OPENGPT_TCP_KEEPALIVE=] [default: 5]
      --tls-cert <TLS_CERT>
          TLS certificate file path [env: OPENGPT_TLS_CERT=]
      --tls-key <TLS_KEY>
          TLS private key file path (EC/PKCS8/RSA) [env: OPENGPT_TLS_KEY=]
  -S, --sign-secret-key <SIGN_SECRET_KEY>
          Enable url signature (signature secret key) [env: OPENGPT_SIGNATURE=]
  -T, --tb-enable
          Enable token bucket flow limitation [env: OPENGPT_TB_ENABLE=]
      --tb-store-strategy <TB_STORE_STRATEGY>
          Token bucket store strategy (mem/redis) [env: OPENGPT_TB_STORE_STRATEGY=] [default: mem]
      --tb-capacity <TB_CAPACITY>
          Token bucket capacity [env: OPENGPT_TB_CAPACITY=] [default: 60]
      --tb-fill-rate <TB_FILL_RATE>
          Token bucket fill rate [env: OPENGPT_TB_FILL_RATE=] [default: 1]
      --tb-expired <TB_EXPIRED>
          Token bucket expired (second) [env: OPENGPT_TB_EXPIRED=] [default: 86400]
  -h, --help
          Print help
```

### Compile

- Linux compile, Ubuntu machine for example:

```shell

sudo apt update -y && sudo apt install rename

# Native compilation
git clone https://github.com/gngpp/opengpt.git && cd opengpt
./build.sh

# Cross-platform compilation, relying on docker (if you can solve cross-platform compilation dependencies on your own)
./build_cross.sh # Default using docker build linux/windows platform 
os=macos ./build_cross.sh # The MacOS platform is built on MacOS by default

# Compile a single platform binary, take aarch64-unknown-linux-musl as an example: 
docker run --rm -it \
  -v $(pwd):/home/rust/src \
  -v $HOME/.cargo/registry:/root/.cargo/registry \  # If you want to use local cache
  -v $HOME/.cargo/git:/root/.cargo/git \  # If you want to use local cache
  ghcr.io/gngpp/opengpt-builder:aarch64-unknown-linux-musl \
  cargo build --release
```

- OpenWrt compile

```shell
cd package
svn co https://github.com/gngpp/opengpt/trunk/openwrt
cd -
make menuconfig # choose LUCI->Applications->luci-app-opengpt  
make V=s
```

### Preview

![img0](./doc/img/img0.png)
![img1](./doc/img/img1.png)
![img2](./doc/img/img2.png)

### Reference

- <https://github.com/tjardoo/openai-client>
- <https://github.com/jpopesculian/reqwest-eventsource>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [v0.2.1](https://github.com/gngpp/opengpt/releases/tag/v0.2.1) - 2023-07-21

<small>[Compare with v0.2.0](https://github.com/gngpp/opengpt/compare/v0.2.0...v0.2.1)</small>

## [v0.2.0](https://github.com/gngpp/opengpt/releases/tag/v0.2.0) - 2023-07-18

<small>[Compare with v0.1.9](https://github.com/gngpp/opengpt/compare/v0.1.9...v0.2.0)</small>

## [v0.1.9](https://github.com/gngpp/opengpt/releases/tag/v0.1.9) - 2023-07-18

<small>[Compare with v0.1.8](https://github.com/gngpp/opengpt/compare/v0.1.8...v0.1.9)</small>

## [v0.1.8](https://github.com/gngpp/opengpt/releases/tag/v0.1.8) - 2023-07-17

<small>[Compare with v0.1.7](https://github.com/gngpp/opengpt/compare/v0.1.7...v0.1.8)</small>

### Added

- Add CHANGELOG.md ([3c25f21](https://github.com/gngpp/opengpt/commit/3c25f21ce9946746e0d2d941a408f58cd5e148fe) by gngpp).

### Fixed

- fix windows build ([9ef8bce](https://github.com/gngpp/opengpt/commit/9ef8bce747c932abc191ebf30849dc299c54afb2) by gngpp).

### Changed

- change template Help & FQA link ([a4c8b42](https://github.com/gngpp/opengpt/commit/a4c8b429f15f4bf7c55a8306bd2bd9057daceea8) by gngpp).

## [v0.1.7](https://github.com/gngpp/opengpt/releases/tag/v0.1.7) - 2023-07-15

<small>[Compare with v0.1.6](https://github.com/gngpp/opengpt/compare/v0.1.6...v0.1.7)</small>

## [v0.1.6](https://github.com/gngpp/opengpt/releases/tag/v0.1.6) - 2023-07-15

<small>[Compare with v0.1.5](https://github.com/gngpp/opengpt/compare/v0.1.5...v0.1.6)</small>

### Added

- Adds support for ALL_PROXY ([c9b1baa](https://github.com/gngpp/opengpt/commit/c9b1baa093bc3040e6d3a960d9f04c67ffa50f88) by gngpp).

## [v0.1.5](https://github.com/gngpp/opengpt/releases/tag/v0.1.5) - 2023-07-12

<small>[Compare with v0.1.4](https://github.com/gngpp/opengpt/compare/v0.1.4...v0.1.5)</small>

### Added

- Add x86_64-pc-windows-msvc target support ([353cc43](https://github.com/gngpp/opengpt/commit/353cc435f506e4ebdc2786d7f969ab8ae089bc94) by gngpp).

## [v0.1.4](https://github.com/gngpp/opengpt/releases/tag/v0.1.4) - 2023-07-11

<small>[Compare with v0.1.3](https://github.com/gngpp/opengpt/compare/v0.1.3...v0.1.4)</small>

## [v0.1.3](https://github.com/gngpp/opengpt/releases/tag/v0.1.3) - 2023-07-10

<small>[Compare with v0.1.2](https://github.com/gngpp/opengpt/compare/v0.1.2...v0.1.3)</small>

## [v0.1.2](https://github.com/gngpp/opengpt/releases/tag/v0.1.2) - 2023-07-10

<small>[Compare with v0.1.1](https://github.com/gngpp/opengpt/compare/v0.1.1...v0.1.2)</small>

### Added

- Add openwrt action ([edde18a](https://github.com/gngpp/opengpt/commit/edde18ae65f1961228f3bbc3bf438274e47d29a2) by gngpp).
- Add openwrt support ([71fdd5a](https://github.com/gngpp/opengpt/commit/71fdd5a6c79beaf0302959d5bbe1218b49afe988) by gngpp).

## [v0.1.1](https://github.com/gngpp/opengpt/releases/tag/v0.1.1) - 2023-07-08

<small>[Compare with v0.1.0](https://github.com/gngpp/opengpt/compare/v0.1.0...v0.1.1)</small>

## [v0.1.0](https://github.com/gngpp/opengpt/releases/tag/v0.1.0) - 2023-07-08

<small>[Compare with first commit](https://github.com/gngpp/opengpt/compare/fb2d098c05b60987ca7874c72a094ad9f5ab474a...v0.1.0)</small>

### Added

- Add cargo registry ([9bc5886](https://github.com/gngpp/opengpt/commit/9bc588637e1b849ff27b50335a826310441247fb) by gngpp).
- Add docker builder image ([1ace0f5](https://github.com/gngpp/opengpt/commit/1ace0f5c78e741145752097cac011bb65206aa33) by gngpp).
- Add request eventsource mod ([bd5bbe0](https://github.com/gngpp/opengpt/commit/bd5bbe0a426f05337323fd5d6be5d109edba2c6a) by gngpp).
- Add platform api client ([3f077ed](https://github.com/gngpp/opengpt/commit/3f077ed6923d8bab9d08d5a8ce4b6141bb53632b) by gngpp).
- Add http rest doc ([43994c6](https://github.com/gngpp/opengpt/commit/43994c61eb748e0808660f23d95ecd5a3c6cf5bd) by gngpp).
- Add billing credit_grants api ([fcfd054](https://github.com/gngpp/opengpt/commit/fcfd054387f8da9653463a31fa9edc1eab938a71) by gngpp).
- Add billing_usage api ([b12f31f](https://github.com/gngpp/opengpt/commit/b12f31fd6f293f0077d42e9c5c23091e94e151ca) by gngpp).
- Add new api ([f613183](https://github.com/gngpp/opengpt/commit/f6131830d14c782826ced5ecd6f0a523a0e801d1) by gngpp).
- Add ui mod ([971d259](https://github.com/gngpp/opengpt/commit/971d2594b04869d27d83851632bf982fc6cde215) by gngpp).

### Fixed

- fix decode escape convert ([51e5bd6](https://github.com/gngpp/opengpt/commit/51e5bd6b09e7157cbf94d62a72acde710343719f) by gngpp).
- fix ios device api aceess ([f3b1e33](https://github.com/gngpp/opengpt/commit/f3b1e33e7fe445401edf6e586199c336bd089e3d) by gngpp).

### Removed

- remove commit ([4fff81a](https://github.com/gngpp/opengpt/commit/4fff81a78aa8943c7e30c5f90fe3e5ffea872ab9) by gngpp).



src/args.rs
use anyhow::Context;
use clap::{Args, Parser, Subcommand};
use openai::serve::tokenbucket;
use serde::{Deserialize, Serialize};
use std::{path::PathBuf, str::FromStr};

#[derive(Parser)]
#[clap(author, version, about)]
pub(super) struct Opt {
    #[clap(subcommand)]
    pub(super) command: Option<SubCommands>,
    /// Log level (info/debug/warn/trace/error)
    #[clap(
        short = 'L',
        long,
        global = true,
        env = "OPENGPT_LOG_LEVEL",
        default_value = "info"
    )]
    pub(super) level: String,
}

#[derive(Subcommand)]
pub(super) enum ServeSubcommand {
    /// Run the HTTP server
    Run(ServeArgs),
    /// Stop the HTTP server daemon
    #[cfg(target_family = "unix")]
    Stop,
    /// Start the HTTP server daemon
    #[cfg(target_family = "unix")]
    Start(ServeArgs),
    /// Restart the HTTP server daemon
    #[cfg(target_family = "unix")]
    Restart(ServeArgs),
    /// Status of the Http server daemon process
    #[cfg(target_family = "unix")]
    Status,
    /// Show the Http server daemon log
    #[cfg(target_family = "unix")]
    Log,
    /// Generate config template file (toml format file)
    GT {
        /// Overwrite existing configuration file
        #[clap(short, long, env = "OPENGPT_SERVE_GT_COVER")]
        cover: bool,
        /// Configuration template output to file (toml format file)
        #[clap(short, long, env = "OPENGPT_SERVE_GT_OUT")]
        out: Option<PathBuf>,
    },
}

#[derive(Args, Debug, Default, Serialize, Deserialize)]
pub(super) struct ServeArgs {
    /// Configuration file path (toml format file)
    #[clap(short = 'C', long, env = "OPENGPT_CONFIG", value_parser = parse_config)]
    pub(super) config: Option<PathBuf>,
    /// Server Listen host
    #[clap(short = 'H', long, env = "OPENGPT_HOST", default_value = "0.0.0.0", value_parser = parse_host)]
    pub(super) host: Option<std::net::IpAddr>,
    /// Server Listen port
    #[clap(short = 'P', long, env = "OPENGPT_PORT", default_value = "7999", value_parser = parse_port_in_range)]
    pub(super) port: Option<u16>,
    /// Server worker-pool size (Recommended number of CPU cores)
    #[clap(short = 'W', long, env = "OPENGPT_WORKERS", default_value = "1")]
    pub(super) workers: usize,
    /// Server proxies pool, example: protocol://user:pass@ip:port
    #[clap(long, env = "OPENGPT_PROXY", value_parser = parse_url)]
    pub(super) proxies: Option<std::vec::Vec<String>>,
    /// Client timeout (secends)
    #[clap(long, env = "OPENGPT_TIMEOUT", default_value = "600")]
    pub(super) timeout: usize,
    /// Client connect timeout (secends)
    #[clap(long, env = "OPENGPT_CONNECT_TIMEOUT", default_value = "60")]
    pub(super) connect_timeout: usize,
    /// TCP keepalive (secends)
    #[clap(long, env = "OPENGPT_TCP_KEEPALIVE", default_value = "60")]
    pub(super) tcp_keepalive: usize,
    /// TLS certificate file path
    #[clap(long, env = "OPENGPT_TLS_CERT", requires = "tls_key")]
    pub(super) tls_cert: Option<PathBuf>,
    /// TLS private key file path (EC/PKCS8/RSA)
    #[clap(long, env = "OPENGPT_TLS_KEY", requires = "tls_cert")]
    pub(super) tls_key: Option<PathBuf>,
    /// Web UI api prefix
    #[clap(long, env = "OPENGPT_UI_API_PREFIX", value_parser = parse_url)]
    pub(super) api_prefix: Option<String>,
    /// Enable url signature (signature secret key)
    #[clap(short = 'S', long, env = "OPENGPT_SIGNATURE")]
    #[cfg(feature = "sign")]
    pub(super) sign_secret_key: Option<String>,
    /// Enable token bucket flow limitation
    #[clap(short = 'T', long, env = "OPENGPT_TB_ENABLE")]
    #[cfg(feature = "limit")]
    pub(super) tb_enable: bool,
    /// Token bucket store strategy (mem/redis)
    #[clap(
        long,
        env = "OPENGPT_TB_STORE_STRATEGY",
        default_value = "mem",
        requires = "tb_enable"
    )]
    #[cfg(feature = "limit")]
    pub(super) tb_store_strategy: tokenbucket::Strategy,
    /// Token bucket redis url(support cluster), example: redis://user:pass@ip:port
    #[clap(
        long,
        env = "OPENGPT_TB_REDIS_URL",
        default_value = "redis://127.0.0.1:6379",
        requires = "tb_enable",
        value_parser = parse_url
    )]
    #[cfg(feature = "limit")]
    pub(super) tb_redis_url: std::vec::Vec<String>,
    /// Token bucket capacity
    #[clap(
        long,
        env = "OPENGPT_TB_CAPACITY",
        default_value = "60",
        requires = "tb_enable"
    )]
    #[cfg(feature = "limit")]
    pub(super) tb_capacity: u32,
    /// Token bucket fill rate
    #[clap(
        long,
        env = "OPENGPT_TB_FILL_RATE",
        default_value = "1",
        requires = "tb_enable"
    )]
    #[cfg(feature = "limit")]
    pub(super) tb_fill_rate: u32,
    /// Token bucket expired (second)
    #[clap(
        long,
        env = "OPENGPT_TB_EXPIRED",
        default_value = "86400",
        requires = "tb_enable"
    )]
    #[cfg(feature = "limit")]
    pub(super) tb_expired: u32,

    /// Cloudflare turnstile captcha site key
    #[clap(long, env = "OPENGPT_CF_SITE_KEY", requires = "cf_secret_key")]
    pub(super) cf_site_key: Option<String>,
    /// Cloudflare turnstile captcha secret key
    #[clap(long, env = "OPENGPT_CF_SECRET_KEY", requires = "cf_site_key")]
    pub(super) cf_secret_key: Option<String>,

    /// Disable WebUI
    #[clap(short = 'D', long, env = "OPENGPT_DISABLE_WEBUI")]
    pub(super) disable_webui: bool,
}

#[derive(Subcommand)]
pub(super) enum SubCommands {
    /// Start the http server
    #[clap(subcommand)]
    Serve(ServeSubcommand),
    /// Account configuration settings
    Account,
    /// Configuration Settings
    Config {
        /// Working directory, refresh_token will be stored in there if specified
        #[clap(short, long, env = "OPENGPT_WORKDIR")]
        workdir: Option<PathBuf>,

        /// Unofficial API prefix. Format: https://example.com/backend-api
        #[clap(long, env = "OPENGPT_API")]
        unofficial_api: Option<String>,

        /// Unofficial API http proxy. Format: protocol://user:pass@ip:port
        #[clap(long, env = "OPENGPT_PROXY", value_parser = parse_url)]
        unofficial_proxy: Option<url::Url>,
    },
}

const PORT_RANGE: std::ops::RangeInclusive<usize> = 1024..=65535;

// port range parse
fn parse_port_in_range(s: &str) -> anyhow::Result<u16> {
    let port: usize = s
        .parse()
        .map_err(|_| anyhow::anyhow!(format!("`{}` isn't a port number", s)))?;
    if PORT_RANGE.contains(&port) {
        return Ok(port as u16);
    }
    anyhow::bail!(format!(
        "Port not in range {}-{}",
        PORT_RANGE.start(),
        PORT_RANGE.end()
    ))
}

// address parse
fn parse_host(s: &str) -> anyhow::Result<std::net::IpAddr> {
    let addr = s
        .parse::<std::net::IpAddr>()
        .map_err(|_| anyhow::anyhow!(format!("`{}` isn't a ip address", s)))?;
    Ok(addr)
}

// proxy proto
fn parse_url(s: &str) -> anyhow::Result<Vec<String>> {
    let split = s.split(",");
    let mut proxies: Vec<_> = vec![];
    for ele in split {
        let url = url::Url::parse(ele)
            .context("The Proxy Url format must be `protocol://user:pass@ip:port`")?;
        let protocol = url.scheme().to_string();
        match protocol.as_str() {
            "http" | "https" | "socks5" | "redis" => proxies.push(ele.to_string()),
            _ => anyhow::bail!("Unsupported protocol: {}", protocol),
        };
    }
    Ok(proxies)
}

fn parse_config(s: &str) -> anyhow::Result<PathBuf> {
    let path =
        PathBuf::from_str(s).map_err(|_| anyhow::anyhow!(format!("`{}` isn't a path", s)))?;
    match path.exists() {
        true => Ok(path),
        false => {
            if let Some(parent) = path.parent() {
                parent.exists().then(|| ()).ok_or_else(|| {
                    anyhow::anyhow!(format!("Path {} not exists", parent.display()))
                })?;
            }
            Ok(path)
        }
    }
}
